whelktool/scripts/cleanups/2020/09/lxl-3304-move-bib976.groovy

```diff
@@ -0,0 +1,136 @@
+/**
+ * See LXL-3304 for more information.
+ */
+
+class Script {
+    static PrintWriter movedToInstance
+    static PrintWriter movedToClassification
+    static PrintWriter removed
+    static PrintWriter equivalentCodes
+    static PrintWriter errors
+}
+
+Script.movedToInstance = getReportWriter("moved-to-instance.txt")
+Script.movedToClassification = getReportWriter("moved-to-classification.txt")
+Script.removed = getReportWriter("removed.txt")
+Script.equivalentCodes = getReportWriter("equivalent-codes.txt")
+Script.errors = getReportWriter("errors.txt")
+
+selectByCollection('bib') { bib ->
+    try {
+        process(bib)
+    }
+    catch(Exception e) {
+        Script.errors.println("${bib.doc.shortId} $e")
+        e.printStackTrace(Script.errors)
+    }
+}
+
+void process(bib) {
+    def instance = bib.graph[1]
+    def work = instance['instanceOf']
+
+    if (!work) {
+        return
+    }
+
+    def bib976 = asList(work['marc:hasBib976'])
+
+    if (!bib976) {
+        return
+    }
+
+    def (code, noCode) = bib976.split { it['marc:bib976-a'] }
+    def bib81 = sab(work)
+
+    handleWithSabCode(bib, work, bib81, code)
+    handleWithoutSabCode(bib, work, instance, noCode)
+}
+
+void handleWithSabCode(bib, work, bib084, bib976) {
+    def (in084, notIn084) = bib976.split {
+        bib084.findAll { sab ->
+            def codes = asList(sab)
+            return codes.any { code ->
+                haveEquivalentCodes(bib, code, it)
+            }
+        }
+    }
+
+    in084.each {
+        bib976.remove(it)
+        print(['bib976-a: ' + it['marc:bib976-a']],
+                bib,
+                Script.removed)
+        bib.scheduleSave()
+    }
+
+    notIn084.each {
+        work['classification'] = work['classification'] ?: []
+        work['classification'].add(createClassification(it['marc:bib976-a'], it['marc:bib976-i2'] ?: 'n/a'))
+        print(['bib976-a: ' + it['marc:bib976-a']],
+                bib,
+                Script.movedToClassification)
+
+        bib.scheduleSave()
+    }
+}
+
+void handleWithoutSabCode(bib, work, instance, bib976) {
+    bib976.each {
+        copyToInstance(instance, it)
+    }
+    work.remove('marc:hasBib976')
+    print(["bib976-a:" + bib976['marc:bib976-a']],
+            bib,
+            Script.movedToInstance)
+    bib.scheduleSave()
+}
+
+void copyToInstance(instance, it) {
+    instance['marc:hasBib976'] = instance['marc:hasBib976'] ?: []
+    instance['marc:hasBib976'].add(it)
+}
+
+boolean haveEquivalentCodes(bib, code084, bib976) {
+    def code976 = bib976['marc:bib976-a']
+    if (code084?.startsWith(code976) ||
+            (code084 == "He=c" && code976 == "Hcee") ||
+            (code084 == "Hcee" && code976 == "He=c")) {
```

Comment:
Maybe we should remove these since both forms are used by different libraries.

Reply:
haveEquivalentCodes => sorted into in084 => removed
The code logging is just an extra info log on top of that.

Reply:
Or what do you mean by 'remove these'?

Reply:
Att He=c inte ska vara lika med Hcee. Det betyder ju samma men det är ok att båda finns i klassifikation eftersom olika bibliotek använder olika.

Reply:
Ja kanske det. Har inte varit med i diskussionen om hur en fin Classification ska se ut, men känns ju lite onödigt att ha flera koder som betyder samma sak å andra sidan. Sen vet jag inte hur det påverkar de enskilda biblioteken om deras föredragna variant skulle försvinna.

Reply:
Sen tycker jag iofs att städning av Classification kan hållas separerat från flyttning av 976. Det kan väl redan idag finnas He=c och Hcee _inom_ klassifikation.